pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -734,6 +737,147 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
 		})
 }
 
+func runForOnDeletePodScenarios(t *testing.T,
+	testFn func(*testing.T, *apps.ControllerRevision, *apps.StatefulSet, []*v1.Pod, []*apps.ControllerRevision),
+) {
+	type testCase struct {
+		name      string
+		expected  *apps.ControllerRevision
+		set       *apps.StatefulSet
+		pods      []*v1.Pod
+		revisions []*apps.ControllerRevision
+	}
+
+	cases := []testCase{}
+
+	//Test case: stable statefulset
+	set := newStatefulSet(3)
+	set.Spec.UpdateStrategy = apps.StatefulSetUpdateStrategy{
+		Type: apps.OnDeleteStatefulSetStrategyType,
+	}
+	set.Status.CollisionCount = new(int32)
+	rev0 := newRevisionOrDie(set, 1)
+	set.Status.CurrentRevision = rev0.Name
+	set.Status.UpdateRevision = rev0.Name
+	pod0rev0 := newStatefulSetPod(set, 0)
+	pod1rev0 := newStatefulSetPod(set, 1)
+	pod2rev0 := newStatefulSetPod(set, 2)
+	setPodRevision(pod0rev0, rev0.Name)
+	setPodRevision(pod1rev0, rev0.Name)
+	setPodRevision(pod2rev0, rev0.Name)
+	cases = append(cases, testCase{
+		name:      "StableRevision",
+		expected:  rev0,
+		set:       set,
+		pods:      []*v1.Pod{pod0rev0, pod1rev0, pod2rev0},
+		revisions: []*apps.ControllerRevision{rev0},
+	})
+
+	//Test case: new revision with one new pod
+	set1 := set.DeepCopy()
+	set1.Spec.Template.Spec.Containers[0].Image = "foo"
+	set1.Status.CurrentRevision = rev0.Name
+	set1.Status.CollisionCount = new(int32)
+	rev1 := newRevisionOrDie(set1, 2)
+	set1.Status.UpdateRevision = rev1.Name
+	pod0rev1 := newStatefulSetPod(set1, 0)
+	setPodRevision(pod0rev1, rev1.Name)
+	cases = append(cases, testCase{
+		name:      "OneNewRevision",
+		expected:  rev0,
+		set:       set1,
+		pods:      []*v1.Pod{pod0rev1, pod1rev0, pod2rev0},
+		revisions: []*apps.ControllerRevision{rev0, rev1},
+	})
+
+	//Test case: pods spread across three revisions
+	set2 := set1.DeepCopy()
+	set2.Spec.Template.Spec.Containers[0].Image = "bar"
+	set2.Status.CurrentRevision = rev0.Name
+	set2.Status.CollisionCount = new(int32)
+	rev2 := newRevisionOrDie(set2, 3)
+	set2.Status.UpdateRevision = rev2.Name
+	pod1rev2 := newStatefulSetPod(set2, 1)
+	setPodRevision(pod1rev2, rev2.Name)
+	cases = append(cases, testCase{
+		name:      "TwoNewRevisions",
+		expected:  rev0,
+		set:       set2,
+		pods:      []*v1.Pod{pod0rev1, pod1rev2, pod2rev0},
+		revisions: []*apps.ControllerRevision{rev0, rev1, rev2},
+	})
+
+	//Test case: remove pod with current revision
+	set3 := set2.DeepCopy()
+	pod2rev2 := newStatefulSetPod(set3, 2)
+	setPodRevision(pod2rev2, rev2.Name)
+	cases = append(cases, testCase{
+		name:      "RemoveOldestRevisionPod",
+		expected:  rev1,
+		set:       set3,
+		pods:      []*v1.Pod{pod0rev1, pod1rev2, pod2rev2},
+		revisions: []*apps.ControllerRevision{rev0, rev1, rev2},
```

Comment:
can we also add tests for
- list of revisions that is not starting from 0
- list is missing revisions
- pods that are not sorted
- pods only have latest revision
- there is no intersection between pods and revisions

Reply:
@atiratree for `list is missing revisions` I assume you're wanting to test where either the CurrentRevision or UpdateRevision is missing from the list of revisions?

Reply:
Also for `list of revisions that is not starting from 0` are you aware of a helper method to generate revisions, or should I just be creating statefulsets with arbitrary changes to generate them?

Reply:
Likewise with `there is no intersection between pods and revisions` this is a situation that can't occur as the code is currently structured.  There will always be an updateRevision